File: Assets/Scripts/processing/Streams.cs
using System;
using System.Collections.Generic;
using System.IO;
using UnityEngine;
using System.Collections.Concurrent;
using System.Linq;

public interface IInput<T> { 
    public void Input(T t, bool block); 
}
public interface IOutput<T> { 
    public T Output(bool block);
}

public interface IStream<T> : IInput<T>, IOutput<T> {}

public class SingleLabelList : List<(DataList dataList,string name)> {}
public class MultiLabelList : List<(DataList dataList,List<string> names)> {}

public class Delta : IInput<SingleLabelList> {
    private Func<string, (IInput<DataList>,bool)> resolver;

    public Delta(Func<string, (IInput<DataList>,bool)> resolver) {
        this.resolver = resolver;
    }
    
    public void Input(SingleLabelList sll, bool block) {
        sll.ForEach(p => {
            var (input, local) = resolver.Invoke(p.name);
            p.dataList.GetState().complete = local;
            input?.Input(p.dataList, block);
        });
    }
}

public class Queue : BlockingCollection<DataList> {}

public class QueueStream : IStream<DataList> {
    private Queue queue = new();

    public QueueStream(Queue queue = null) {
        this.queue = queue ?? new();
    }

    public virtual void Input(DataList dl, bool block) {
        if(block) queue.Add(dl);
        else queue.TryAdd(dl);
    }

    public virtual DataList Output(bool block) {
        if(block) return queue.Take();
        bool took = queue.TryTake(out DataList dl);
        if(took) return dl;
        else throw new SkipExeception();
    }
}

public class SocketStream : IStream<DataList> {
    public Peer peer;

    public SocketStream(Peer peer) {
        this.peer = peer;
    }

    public void Input(DataList dl, bool block) {
        if(!block) throw new Exception("SocketStream must block");
        if(peer.CanSend()) peer.SendDataList(dl);
    }

    public DataList Output(bool block) {
        if(!block) throw new Exception("SocketStream must block");
        return peer.CanReceive() ? peer.ReceiveDataList() : null;
    }
}

public class NullStream : IStream<object> {
    public void Input(object t, bool block) { }
    public object Output(bool block) { return null; }
}

public class Rapid : IInput<MultiLabelList>, IOutput<DataList> {
    private QueueStream queueStream;

    public Rapid(QueueStream queueStream=null) {
        this.queueStream = queueStream ?? new();
    }

    public void Input(MultiLabelList mll, bool block) {
        mll.ForEach(p => {
            p.dataList.SetSelected(p.names);
            queueStream.Input(p.dataList, block);
        });
    }

    public DataList Output(bool block) {
        return queueStream.Output(block);
    }
}

public class Confluence : QueueStream {
    private Dictionary<int, Group> mainGrouplistDict;
    private Dictionary<int, Dictionary<int,int>> mainStateGroupDict;
    private Dictionary<int, Dictionary<int, DataList>> exampleDict;

    public Confluence(Dictionary<int, Group> confluenceDict, Queue queue=null) {
        mainGrouplistDict = confluenceDict;
        mainStateGroupDict = CreateMainStateGroupDict();
        exampleDict = CreateExampleDict();
    }

    private Dictionary<int, Dictionary<int,int>> CreateMainStateGroupDict() {
        return mainGrouplistDict.ToDictionary(
            pair => pair.Key,
            pair => pair.Value.SelectMany((list, i) => list.Select(s => new{ s, i }))
                              .ToDictionary(p => p.s, p => p.i)
        );
    }

    private Dictionary<int, Dictionary<int, DataList>> CreateExampleDict() {
        return mainGrouplistDict.ToDictionary(
            p => p.Key,
            p => p.Value.SelectMany(l => l).ToDictionary(s => s, _ => (DataList)null)
        );
    }

    private int DLToStateID(DataList dl) { return dl.GetPreviousState().id; }
    private int DLToGroupID(DataList dl) { return mainStateGroupDict[DLToMainID(dl)][DLToStateID(dl)]; }
    private int DLToMainID(DataList dl) { return dl.GetState().id; }
    private (int,int,int) DLToIDs(DataList dl) { return (DLToStateID(dl), DLToMainID(dl), DLToGroupID(dl)); }

    //code below is based on current datalist's ids
    private int mid,sid,gid;
    
    private List<DataList> GetGroupExamples() {
        return mainGrouplistDict[mid][gid].Select(s => exampleDict[mid][s]).ToList();
    }

    private bool GroupSatisfied() {
        return GetGroupExamples().All(dl => dl != null);
    }

    private DataList Merge() {
        var examples  = GetGroupExamples();
        DataList mergedDL = new();
        mergedDL.ChangeState(examples[0].GetState());
        mergedDL.SetPreviousState(gid); //gimmick
        examples.ForEach(e => {
            mergedDL.AddContent(e.GetContent());
            exampleDict[mid][DLToStateID(e)] = null;
        });
        return mergedDL;
    }

    private bool InsertDataList(DataList dl) {
        bool wasEmpty = exampleDict[mid][sid] == null;
        exampleDict[mid][sid] = dl;
        return wasEmpty;
    }

    public override void Input(DataList dl, bool block) {
        if(dl == null) {base.Input(dl, block); return;}
        (sid, mid, gid) = DLToIDs(dl);
        if(InsertDataList(dl) && GroupSatisfied()) base.Input(Merge(), block); 
    }
}

File: Assets/Python/root/processing/streams.py
from typing import Generic, TypeVar, Any, Callable
from queue import Queue

from networking.peers import Peer
from datastructs.datalist import DataList

T = TypeVar('T')
class Input(Generic[T]):
    def input(self, t:T): pass
class Output(Generic[T]):
    def output(self) -> T: pass

class Stream(Input[T], Output[T]):
    pass

SingleLabelList = list[tuple[DataList,str]]
MultiLabelList = list[tuple[DataList,list[str]]]

#routes datalists to different inputs
class Delta(Input[SingleLabelList]):
    def __init__(self, name_to_input: Callable[[str], Input[DataList]]):
        self.name_to_input = name_to_input

    def input(self, sll: SingleLabelList):
        for dl, sn in sll: 
            input = self.name_to_input(sn)
            if input: input.input(dl)

#Simple stream using a queue for datalists
class QueueStream(Stream[DataList]):
    def __init__(self, queue: Queue=None) -> None:
        self.queue = queue or Queue()

    def input(self, d: DataList): 
        self.queue.put(d)

    def output(self) -> DataList:
        return self.queue.get()
    
#Stream implementing a socket endpoint
class SocketStream(Stream[DataList]):
    def __init__(self, peer: Peer) -> None:
        self.peer = peer

    def input(self, d: DataList):
        if not self.peer.can_send(): return
        self.peer.send_datalist(d)

    def output(self) -> DataList:
        if self.peer.can_receive():
            return self.peer.receive_datalist()
        else: return None

#Stream for returning nothing important
class NullStream(Stream[Any]):
    def input(self, t: Any):
        pass

    def output(self) -> Any:
        return None
    
#takes a list of Datalists and labels them
class Rapid(Input[MultiLabelList], Output[DataList]):
    def __init__(self, queue: QueueStream=None) -> None:
        self.queue = queue or QueueStream()

    def input(self, mll: MultiLabelList):
        for dl, selected in mll:
            dl.set_selected(selected)
            self.queue.input(dl)

    def output(self) -> DataList:
        return self.queue.output()
    
Group = list[list[int]]
    
#Stream that takes in multiple values and only passes them on if they finish a group, groups can be different depending on when the service is encountered
class Confluence(QueueStream):
    def __init__(self, groups: dict[int, Group], queue: Queue = None) -> None:
        super().__init__(queue)
        self.main_group_list_dict = groups
        self.input_state_group_dict = self.create_input_state_group_dict(groups)
        self.example_dict = {m: {s: None for g in G for s in g} for m,G in groups.items()}

    def create_input_state_group_dict(self, groups: dict[int, Group]):
        return {m: {s: i for i, g in enumerate(G) for s in g} for m,G in groups.items()}
    
    def dl_to_id(self, dl: DataList):
        return dl.get_previous_state().id
    def dl_to_group(self, dl: DataList):
        return self.input_state_group_dict[self.dl_to_main(dl)][self.dl_to_id(dl)]
    def dl_to_main(self, dl: DataList):
        return dl.get_state().id
    def dl_ids(self, dl: DataList):
        return (self.dl_to_id(dl), self.dl_to_main(dl), self.dl_to_group(dl))
    
    def get_group_examples(self) -> list[DataList]:
        return [self.example_dict[self.mid][s] for s in self.main_group_list_dict[self.mid][self.gid]]
    
    def group_satisfied(self):
        return all(self.get_group_examples())
    #merges several datalists into one
    def merge(self):
        examples = self.get_group_examples()
        merged_dl = DataList()
        merged_dl.set_state(examples[0].get_state())
        merged_dl.set_previous_state(self.gid) #only a gimmick
        for e in examples:
            merged_dl.add_contents(e.content)
            self.example_dict[self.mid][self.dl_to_id(e)] = None
        return merged_dl
    
    #inserts a datalist into its storage
    def insert_datalist(self, dl: DataList):
        was_empty = self.example_dict[self.mid][self.sid] is None
        self.example_dict[self.mid][self.sid] = dl
        return was_empty
    
    def input(self, dl: DataList):
        if not isinstance(dl, DataList): 
            super().input(dl)
            return
        self.sid, self.mid, self.gid = self.dl_ids(dl)
        was_empty = self.insert_datalist(dl)
        if not was_empty: return
        if self.group_satisfied():
            super().input(self.merge())